cmd/promtool/backfill_test.go

```diff
@@ -0,0 +1,265 @@
+func queryblock(t testing.TB, q storage.Querier, matchers ...*labels.Matcher) map[string][]tsdbutil.Sample {
+	ss := q.Select(false, nil, matchers...)
+	defer func() {
+		require.NoError(t, q.Close())
+	}()
+
+	result := map[string][]tsdbutil.Sample{}
+	for ss.Next() {
+		series := ss.At()
+
+		samples := []tsdbutil.Sample{}
+		it := series.Iterator()
+		for it.Next() {
+			t, v := it.At()
+			samples = append(samples, backfillSample{t: t, v: v})
+		}
+		require.NoError(t, it.Err())
+
+		if len(samples) == 0 {
+			continue
+		}
+
+		name := series.Labels().String()
+		result[name] = samples
+	}
+	require.NoError(t, ss.Err())
+	require.Equal(t, 0, len(ss.Warnings()))
+
+	return result
+}
+
+func testBlocks(t *testing.T, blockpath string, mint, maxt int64, expectedSeries map[string][]tsdbutil.Sample) {
+	b, err := tsdb.OpenBlock(nil, blockpath, nil)
+	require.NoError(t, err)
+	q, err := tsdb.NewBlockQuerier(b, math.MinInt64, math.MaxInt64)
+	require.NoError(t, err)
+	series := queryblock(t, q, labels.MustNewMatcher(labels.MatchRegexp, "", ".*"))
+	require.Equal(t, expectedSeries, series)
+}
+func TestBackfill(t *testing.T) {
+	tests := []struct {
+		ToParse  string
+		IsOk     bool
+		Expected struct {
+			MinTime int64
+			MaxTime int64
+			Series  map[string][]tsdbutil.Sample
+		}
+	}{
+		{
+			ToParse: `# EOF`,
+			IsOk:    true,
+		},
+		{
+			ToParse: `# HELP http_requests_total The total number of HTTP requests.
+# TYPE http_requests_total counter
+http_requests_total{code="200"} 1021 1565133713989
+http_requests_total{code="400"} 1 1565133713990
+# EOF
+`,
+			IsOk: true,
+			Expected: struct {
+				MinTime int64
+				MaxTime int64
+				Series  map[string][]tsdbutil.Sample
+			}{
+				MinTime: 1565133713989000,
+				MaxTime: 1565133713990000,
+				Series:  map[string][]tsdbutil.Sample{"{__name__=\"http_requests_total\", code=\"200\"}": []tsdbutil.Sample{backfillSample{t: 1565133713989000, v: 1021}}, "{__name__=\"http_requests_total\", code=\"400\"}": []tsdbutil.Sample{backfillSample{t: 1565133713990000, v: 1}}},
+			},
+		},
+		{
+			ToParse: `# HELP http_requests_total The total number of HTTP requests.
+# TYPE http_requests_total counter
+http_requests_total{code="200"} 1022 1565133713989
+http_requests_total{code="400"} 2 1565133713990
+# EOF
+`,
+			IsOk: true,
+			Expected: struct {
+				MinTime int64
+				MaxTime int64
+				Series  map[string][]tsdbutil.Sample
+			}{
+				MinTime: 1565133713989000,
+				MaxTime: 1565133713990000,
+				Series:  map[string][]tsdbutil.Sample{"{__name__=\"http_requests_total\", code=\"200\"}": []tsdbutil.Sample{backfillSample{t: 1565133713989000, v: 1022}}, "{__name__=\"http_requests_total\", code=\"400\"}": []tsdbutil.Sample{backfillSample{t: 1565133713990000, v: 2}}},
+			},
+		},
+		{
+			ToParse: `# HELP http_requests_total The total number of HTTP requests.
+# TYPE http_requests_total counter
+http_requests_total{code="200"} 1023 1395066363000
+http_requests_total{code="400"} 3 1395066363000
+# EOF
```

Comment:
Those first three tests are all testing the same things, aren't they?

What would be interesting is multiple samples with different timestamp for the same series. And that for two cases: Samples all in the same block. And then samples that would end up in different blocks. (I think your testing code so far couldn't cope with the latter as it checks each block separately. That will be a bit tricky to test. Perhaps it will be best to add to the `Expected` part how many blocks to expect and which series data to expect in each block.)